[ParseableInterface] Print conformances inherited via private protos

jrose-apple opened the pull request:
In this code...

```swift
private protocol MyProto: Hashable {}
public struct MyStruct: MyProto {}
```

...being Hashable is part of MyStruct's public API, even though it's not written explicitly. If we're not going to require people to write it explicitly, we need to make sure it gets printed.

rdar://problem/44662501

jrose-apple commented:
@swift-ci Please test

jrose-apple commented on lib/Frontend/ParseableInterfaceSupport.cpp:
```diff
@@ -400,6 +401,134 @@ static void printImports(raw_ostream &out, ModuleDecl *M) {
   }
 }
 
+// FIXME: Copied from ASTPrinter.cpp...
```
I think Harlan's moving these into a header file anyway in another PR. Once that lands, I'll move PrintOptions::printParseableInterfaceFile up to Frontend, and then we'll only need one copy of these helpers.

harlanhaskins commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -4414,12 +4414,18 @@ swift::getInheritedForPrinting(const Decl *decl,
-        if (!shouldPrint(NTD))
-          continue;
+    if (auto ty = TL.getType()) {
+      bool foundUnprintable = ty.findIf([shouldPrint](Type subTy) {
```
Can you factor this out? Especially if it's gonna be reused in #20146.

jrose-apple replied:
It's not exactly the same query, though: the `shouldPrint` here is an argument to the function the way ASTPrinter is currently written. I suppose Type itself could have an overload of `findIf` that passes GenericTypeDecls.

harlanhaskins commented on lib/Frontend/ParseableInterfaceSupport.cpp:
```diff
@@ -400,6 +401,134 @@ static void printImports(raw_ostream &out, ModuleDecl *M) {
+  return scope.isPublic();
+}
+
+static bool isPublicOrUsableFromInline(Type ty) {
```
Ah! Okay, yes, this copy should probably make its way into the header introduced in #20004 (or whichever PR I land that contains just the cleanup)

harlanhaskins commented on lib/Frontend/ParseableInterfaceSupport.cpp:
```diff
@@ -400,6 +401,134 @@ static void printImports(raw_ostream &out, ModuleDecl *M) {
+    }
+
+    void
+    printSynthesizedExtensionIfNeeded(raw_ostream &out,
```
Can you add a comment here explaining the motivation for why we're printing the empty extensions after the types? I think it's useful to mention that we are printing conformances to public types that are hidden by conformances to private types, and that they're all satisfied by the requirements that have already been printed.

jrose-apple replied:
Oops. Yes, this whole scheme deserves commenting.

jrose-apple replied:
It kind of stinks to print the extensions at the end of the file rather than next to the type, but I don't see a good way around that.

harlanhaskins commented on test/ParseableInterface/conformances.swift:
```diff
@@ -11,10 +11,69 @@ public protocol SimpleProto {
 } // CHECK: {{^}$}}
+
+
+public protocol PublicProto {}
```
Seems the `CHECK`s are missing here

jrose-apple replied:
…right. I was manually inspecting the output but didn't actually write down the CHECK lines I wanted.

jrose-apple committed:
[ParseableInterface] Print protocols inherited through private protos

In this code:

    private protocol MyProto: Hashable {}
    public struct MyStruct: MyProto {}

Being Hashable is part of MyStruct's public API, even though it's not
written explicitly. If we're not going to require people to write it
explicitly, we need to make sure it gets printed.

rdar://problem/44662501

jrose-apple commented:
Most concerns addressed.

@swift-ci Please smoke test

slavapestov commented on lib/Frontend/ParseableInterfaceSupport.cpp:
```diff
@@ -400,6 +401,157 @@ static void printImports(raw_ostream &out, ModuleDecl *M) {
+      ExistentialLayout layout = inheritedTy->getExistentialLayout();
+      for (ProtocolType *protoTy : layout.getProtocols())
+        whichProtocols.push_back(protoTy->getDecl());
+      // FIXME: This ignores layout constraints, but currently we don't support
```
You're ignoring superclass constraints too.

slavapestov reviewed (commented):
Can you check that it does the right thing with subclass existentials? Eg if P is a private protocol and C is a public class, then I would expect `protocol Q : P & C` to print as `protocol Q : C`.

slavapestov commented:
Wait, the example I gave doesn't make sense, but you should still make sure it does something reasonable with subclass existentials.

jrose-apple commented:
Ooh, you're right, I totally forgot to test those. It's okay to ignore superclass constraints because we're printing the class itself, but…

…oh, no, that's not correct; we allow specifying the superclass like that, don't we. Yikes. Okay, I may do that in a follow-up PR.

slavapestov commented:
@jrose-apple It looks like we allow this:

```
public class Base {}
public protocol P2 {}

public class Derived2 : P2 & Base {}
```

But not this:

```
public class Base {}
internal protocol P2 {}

public class Derived2 : P2 & Base {}
```

Which is a bit weird, but I don't think we intended for subclass existentials to work there at all. So it might be that there's nothing to test.

jrose-apple commented:
I think we allow that for people who make typealiases of subclass existentials, but if we already check them for access then I'm happy. Added two more tests.

@swift-ci Please smoke test

jrose-apple commented:
@slavapestov @harlanhaskins Ready to go?
